### DISCUSSION TITLE
boid_flockers example problem

### DISCUSSION THREAD

**Discussion by Apalms** [mesa/mesa#1311] [Q&A]
Why only the boid flockers example does not work with visualization?

**Comment by rht** [comment]
Can you be more specific?

**Comment by Apalms** [comment]
when run the run.py , in the browser appears only this
[cid:1f131ba0-5b06-47f3-9643-1195ff810f22]
________________________________

**Reply by rht** [reply]
Not sure what that is. A screenshot of what it looks like in your browser would be better.

**Reply by clarkpetri** [reply]
I'm having the same issue and I think I can convey it more clearly. From the boids example, I have the run.py file in a folder with a boid_flockers folder containing the remaining files: boid.py, model.py, server.py, simple_continuous_canvas.js, and SimpleContinuousModule.py. When I execute run.py from the terminal I receive the following error:

![Screen Shot 2022-07-13 at 9 15 21 AM](https://user-images.githubusercontent.com/99350552/178742388-3e0dbd2e-d2d1-4fe5-a37c-afa649881a32.png)

All boids files downloaded from git today, 13 July 2022.

The server and visualization sides of MESA remain where I'm weakest so any assistance is appreciated. I'm trying to move a CA model in MESA to continuous space with transfer entropy, so exploration of boids here is helpful. Thanks!

**Reply by rht** [reply]
That line 3 looks outdated. It should have been
`class SimpleCanvas(mesa.visualization.VisualizationElement):`. Are you sure you have pulled from the main branch instead of the deprecated master branch?

**Reply by clarkpetri** [reply]
Apologies, in my somewhat bumbling troubleshooting I had removed the .mesa in an attempt to understand things but that terminal screenshot was from said change. I did download from the main branch originally and with the code as you lay out I receive the following (which I believe gets to OP's original issue)

![Screen Shot 2022-07-13 at 10 34 40 AM](https://user-images.githubusercontent.com/99350552/178760455-96733590-feb9-417e-8833-d026ce91697a.png)

**Reply by rht** [reply]
That error means your Mesa is outdated.

**Reply by clarkpetri** [reply]
Well....there we go.

[Feeling sheepish](https://youtu.be/k1rxlem25ds?t=17) for missing something so obvious.

**Comment by Apalms** [comment]
The screenshot that i sent you is a screenshot of how it look in my browser.

________________________________

**Reply by rht** [reply]
There is no screenshot in any of your messages. If there were, I wouldn't have asked for one.

**Comment by Apalms** [comment]
I'm sorry. Now i believe that there is a screenshot.
________________________________

**Comment by Apalms** [comment]
<img width="953" alt="boids" src="https://user-images.githubusercontent.com/74879906/168548568-bda8f3be-f9c5-4b5c-b872-d9b80e4c2fbb.png">

**Reply by rht** [reply]
1. Can you go to your browser console, (ctrl-shift-i in Chrome/Firefox) and take screenshot if there is any error message?
2. Can you find and paste any strange messages you'd find in the terminal console where you do the `python run.py`?

**Reply by Apalms** [reply]
1. 
![image](https://user-images.githubusercontent.com/74879906/168554876-cfab1a56-57a1-464d-ab21-7095c623629b.png)
2.
WARNING:tornado.access:404 GET /local/boid_flockers/simple_continuous_canvas.js (127.0.0.1) 7.98ms
Socket opened!
{"type":"reset"}
WARNING:tornado.access:404 GET /favicon.ico (127.0.0.1) 2.99ms
WARNING:tornado.access:404 GET /static/external/bootstrap-slider-11.0.2/dist/css/bootstrap-slider.min.css.map (127.0.0.1) 2.99ms

**Reply by rht** [reply]
Have you made sure that you are running the `run.py` from inside the boid_flockers folder, instead the root of the Mesa Git folder?

**Reply by Apalms** [reply]
![image](https://user-images.githubusercontent.com/74879906/168572901-9eb4ba62-c834-4ece-a654-93ba37d545c6.png)
i running the run.py as shown in the screenshot

**Reply by rht** [reply]
You can try both answers in https://stackoverflow.com/questions/67785598/how-to-change-vs-code-working-directory-to-the-running-file-directory.

**Comment by Apalms** [comment]
Unfortunately is not working

**Reply by rht** [reply]
You have to describe what you did and the outcome, otherwise I can't debug.

**Comment by Apalms** [comment]
I installed mesa , after i pip install l -e git+https://github.com/projectmesa/mesa#egg=mesa and so the folder of github appeared in VS code. After that i open run.py from boid_flockers file and then i pressed run to execute the program

**Reply by rht** [reply]
And I have already told you this:
> You can try both answers in https://stackoverflow.com/questions/67785598/how-to-change-vs-code-working-directory-to-the-running-file-directory.

What was the outcome?

**Reply by Apalms** [reply]
the outcome was the same

**Reply by rht** [reply]
Can you also show the server log when trying the SO answers? Please always provide the log, because it is needed for debugging.

**Reply by rht** [reply]
When trying the SO answers, make sure to first undo the `os.path.dirname(__file__)` code change.

**Reply by rht** [reply]
I'm waiting for your finding here. Can you answer the question?

**Comment by rht** [comment]
Can you try this: in SimpleContinuousModel.py, add `import os` to the beginning of the file, then in the same file, replace
```python
local_includes = ["boid_flockers/simple_continuous_canvas.js"]
```
with
```python
local_includes = [os.path.dirname(__file__) + "/simple_continuous_canvas.js"]
```

**Comment by Apalms** [comment]
![image](https://user-images.githubusercontent.com/74879906/168771547-eb7b5eba-f993-445d-b4fa-191600ea8ab3.png)
the outcome is the same

**Reply by rht** [reply]
Please paste the server message like this one.
```
WARNING:tornado.access:404 GET /local/boid_flockers/simple_continuous_canvas.js (127.0.0.1) 7.98ms
Socket opened!
{"type":"reset"}
WARNING:tornado.access:404 GET /favicon.ico (127.0.0.1) 2.99ms
WARNING:tornado.access:404 GET /static/external/bootstrap-slider-11.0.2/dist/css/bootstrap-slider.min.css.map (127.0.0.1) 2.99ms
```

**Reply by Apalms** [reply]
WARNING:tornado.general:403 GET /local/c:/Users/Andreas/Documents/UNI/10o%20semester/thesis/implementations/src/mesa/examples/boid_flockers/boid_flockers/simple_continuous_canvas.js (127.0.0.1): c:\Users\Andreas\Documents\UNI\10o semester\thesis\implementations\src\mesa\examples\boid_flockers\boid_flockers\simple_continuous_canvas.js is not in root static directory  
WARNING:tornado.access:403 GET /local/c:/Users/Andreas/Documents/UNI/10o%20semester/thesis/implementations/src/mesa/examples/boid_flockers/boid_flockers/simple_continuous_canvas.js (127.0.0.1) 11.97ms
Socket opened!
{"type":"reset"}
WARNING:tornado.access:404 GET /favicon.ico (127.0.0.1) 1.99ms

**Reply by Apalms** [reply]
i believe the problem is in space.py because appears 2 problems
![image](https://user-images.githubusercontent.com/74879906/168783290-b33f7510-1297-45b2-bd93-25094391ed6f.png)
